browser/net/brave_proxying_url_loader_factory.cc

```diff
@@ -0,0 +1,969 @@
+      frame_tree_node_id_(frame_tree_node_id),
+      url_loader_header_client_binding_(this),
+      request_id_generator_(request_id_generator),
+      disconnect_callback_(std::move(on_disconnect)),
+      weak_factory_(this) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
+  DCHECK(proxy_bindings_.empty());
+  DCHECK(!target_factory_.is_bound());
+
+  target_factory_.Bind(std::move(target_factory));
+  target_factory_.set_connection_error_handler(
+      base::BindOnce(&BraveProxyingURLLoaderFactory::OnTargetFactoryError,
+                     base::Unretained(this)));
+
+  proxy_bindings_.AddBinding(this, std::move(loader_request));
+  proxy_bindings_.set_connection_error_handler(
+      base::BindRepeating(&BraveProxyingURLLoaderFactory::OnProxyBindingError,
+                          base::Unretained(this)));
+
+  if (header_client_request)
+    url_loader_header_client_binding_.Bind(std::move(header_client_request));
+}
+
+BraveProxyingURLLoaderFactory::~BraveProxyingURLLoaderFactory() = default;
+
+// static
+bool BraveProxyingURLLoaderFactory::MaybeProxyRequest(
+    content::BrowserContext* browser_context,
+    content::RenderFrameHost* render_frame_host,
+    int render_process_id,
+    const url::Origin& request_initiator,
+    network::mojom::URLLoaderFactoryRequest* factory_request,
+    network::mojom::TrustedURLLoaderHeaderClientPtrInfo* header_client) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  auto proxied_request = std::move(*factory_request);
+  network::mojom::URLLoaderFactoryPtrInfo target_factory_info;
+  *factory_request = mojo::MakeRequest(&target_factory_info);
+
+  network::mojom::TrustedURLLoaderHeaderClientRequest header_client_request;
+  if (header_client)
+    header_client_request = mojo::MakeRequest(header_client);
+  base::PostTaskWithTraits(
+      FROM_HERE, {content::BrowserThread::IO},
+      base::BindOnce(
+          &ResourceContextData::StartProxying,
+          browser_context->GetResourceContext(), render_process_id,
+          render_frame_host ? render_frame_host->GetFrameTreeNodeId() : 0,
+          std::move(proxied_request), std::move(target_factory_info),
+          std::move(header_client_request)));
+  return true;
+}
+
+void BraveProxyingURLLoaderFactory::CreateLoaderAndStart(
+    network::mojom::URLLoaderRequest loader_request,
+    int32_t routing_id,
+    int32_t request_id,
+    uint32_t options,
+    const network::ResourceRequest& request,
+    network::mojom::URLLoaderClientPtr client,
+    const net::MutableNetworkTrafficAnnotationTag& traffic_annotation) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
+
+  // The request ID doesn't really matter in the Network Service path. It just
+  // needs to be unique per-BrowserContext so extensions can make sense of it.
+  // Note that |network_service_request_id_| by contrast is not necessarily
+  // unique, so we don't use it for identity here.
+  const uint64_t web_request_id = request_id_generator_->Generate();
```

Comment:
I don't think we need this, it's for extensions

Reply:
we use it as an ID in `BraveRequestHandler`

Reply:
maybe change the name then to avoid confusion with the web_request_api? `brave_request_id`?